relentless/optimize/method.py

```diff
@@ -0,0 +1,193 @@
+
+.. autosummary::
+    :nosignatures:
+
+    Optimizer
+    SteepestDescent
+
+.. autoclass:: Optimizer
+    :member-order: bysource
+    :members: optimize,
+        has_converged,
+        abs_tol
+
+.. autoclass:: SteepestDescent
+    :member-order: bysource
+    :members: optimize,
+        step_size,
+        max_iter
+
+"""
+import abc
+
+from .objective import ObjectiveFunction
+
+class Optimizer(abc.ABC):
+    """Abstract base class for optimization algorithm.
+
+    A :py:class:`Optimizer` defines the optimization algorithm with specified parameters.
+
+    Parameters
+    ----------
+    abs_tol : float or dict
+        The absolute tolerance or tolerances (keyed on the :py:class:`ObjectiveFunction`
+        design variables).
+
+    """
+    def __init__(self, abs_tol):
+        self.abs_tol = abs_tol
+
+    @abc.abstractmethod
+    def optimize(self, objective):
+        """Adjusts the given objective function until the convergence criteria
+        (defined by :py:meth:`has_converged()`) is satisfied.
+
+        Parameters
+        ----------
+        objective : :py:class:`ObjectiveFunction`
+            The objective function to be optimized.
+
+        """
+        pass
+
+    def has_converged(self, result):
+        """Checks if the convergence criteria is satisfied.
+
+        Checks if the absolute value of the gradient for each design variable of the
+        objective function is less than the absolute tolerance for that variable.
+
+        Parameters
+        ----------
+        result : :py:class:`ObjectiveFunctionResult`
+            The computed value of the objective function.
+
+        Returns
+        -------
+        bool
+            ``True`` if the criteria is satisfied, ``False`` otherwise.
+
+        Raises
+        ------
+        KeyError
+            If the absolute tolerance value is not set for all design variables.
+        """
+        for x in result.design_variables:
+            grad = result.gradient(x)
+            try:
+                tol = self.abs_tol[x]
+            except TypeError:
+                tol = self.abs_tol
+            except KeyError:
+                raise KeyError('Absolute tolerance not set for design variable ' + str(x))
+            if abs(grad) > tol:
+                return False
+
+        return True
+
+    @property
+    def abs_tol(self):
+        """float or dict: The absolute tolerance(s). Must be non-negative."""
+        return self._abs_tol
+
+    @abs_tol.setter
+    def abs_tol(self, value):
+        try:
+            abs_tol = dict(value)
+            err = any([tol < 0 for tol in value.values()])
+        except TypeError:
+            abs_tol = value
+            err = value < 0
+        if err:
+            raise ValueError('Absolute tolerances must be non-negative.')
+        else:
+            self._abs_tol = abs_tol
+
+class SteepestDescent(Optimizer):
+    r"""Steepest descent algorithm.
+
```

Comment:
This needs a long description showing the algorithm as an equation with the `.. math::` directive.